build-common/src/org/jetbrains/kotlin/incremental/JavaClassName.kt

```diff
@@ -0,0 +1,210 @@
+
+    /**
+     * Whether this class is an anonymous class.
+     *
+     * Note: Even though an anonymous class has no name in the source code, it always has a (not-null, not-empty) name in the compiled
+     * class. Therefore, [simpleName] is not `null` and not empty even for anonymous classes.
+     * */
+    val isAnonymous: Boolean,
+
+    /** Whether this class is a synthetic class. */
+    val isSynthetic: Boolean
+) : NestedClass(name) {
+
+    init {
+        check(name.startsWith("$outerName\$"))
+    }
+
+    /**
+     * The simple name of this class (e.g., the simple name of "com/example/OuterClass$NestedClass" is "NestedClass", the simple name of
+     * class "com/example/OuterClassWith$Sign$NestedClassWith$Sign" is "NestedClassWith$Sign").
+     *
+     * Note: [simpleName] is not `null` and not empty even for anonymous classes (see [isAnonymous]).
+     */
+    val simpleName: String
+        get() = name.substring("$outerName\$".length)
+}
+
+/** See [JavaClassName]. */
+class LocalClass(name: String) : NestedClass(name)
+
+/**
+ * Computes [ClassId]s of the given Java classes.
+ *
+ * Note that creating a [ClassId] for a nested class will require accessing the outer class for 2 reasons:
+ *   - To disambiguate any '$' characters in the class name (e.g., "com/example/OuterClassWith$Sign$NestedClassWith$Sign").
+ *   - To determine whether a class is a local class (a nested class of a local class is also considered local, see [ClassId]'s kdoc).
+ *
+ * Therefore, outer classes and nested classes must be passed together in one invocation of this method.
+ */
+fun computeJavaClassIds(javaClassNames: List<JavaClassName>): List<ClassId> {
+    val nameToJavaClassName: Map<String, JavaClassName> = javaClassNames.associateBy { it.name }
+    val nameToClassId: MutableMap<String, ClassId?> = nameToJavaClassName.mapValues { null }.toMutableMap()
+
+    fun getOrCreateClassId(className: String): ClassId {
+        val classInfo = nameToJavaClassName[className] ?: error("Class name not found: $className")
+        val computedClassId = nameToClassId[className]
+        if (computedClassId != null) {
+            return computedClassId
+        }
+
+        val packageName = FqName(classInfo.packageName.replace('/', '.'))
+        val classId = when (classInfo) {
+            is TopLevelClass -> {
+                ClassId(packageName, FqName(classInfo.simpleName), /* local */ false)
+            }
+            is NestedNonLocalClass -> {
+                val outerClassId = getOrCreateClassId(classInfo.outerName)
+                val relativeClassName = FqName(outerClassId.relativeClassName.asString() + "." + classInfo.simpleName)
+                // For ClassId, a nested non-local class of a local class is also considered local (see ClassId's kdoc).
+                val isLocal = outerClassId.isLocal
+
+                ClassId(packageName, relativeClassName, /* local */ isLocal)
+            }
+            is LocalClass -> {
```

Comment:
I think the main performance overhead is in parsing all classes (including local classes) to get `JavaClassName`s.

(We can try to avoid parsing local classes by guessing from their paths, but that doesn't sound very safe.)

`ClassId` computation from `JavaClassName` is fast compared to parsing. However, I agree it's still nice if we can avoid it, so I've done that in the next (not-yet-published) PR: https://github.com/google/kotlin/commit/394eb2613cc1806468dfafbc86f0b394de3d1f33#diff-91539f89277edb2d64c5c5d4d67a7c281994c83ffc975b9fdeaa8e539d5a1e44R83